### PR TITLE
Relationships: save blobs in the database instead of in the blob storage 

### PR DESCRIPTION
**Pull request by NikolaVetnic**
# Readiness checklist

-   [ ] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

# Description

Saving relationship change content to blob storage feature substituted with persisting to a database table due to performance issues. A database migration adding the `Content` column to the `"Relationships"."RelationshipChanges"` table was added. Removed saving the content to the blob storage on relationship change creation. Support for "legacy" relationship changes is added by updating the database on project start when blob storage configuration is present.

### CODE REVIEW

File: BuildingBlocks/src/BuildingBlocks.API/Extensions/HostExtensions.cs [hunk 1]
```diff
@@ -69,7 +69,7 @@ public static class HostExtensions
     }
 }
 
-public interface IDbSeeder<in T> where T : DbContext
+public interface IDbSeeder<in T>
```
Comment (tnotheis) [inline BuildingBlocks/src/BuildingBlocks.API/Extensions/HostExtensions.cs:72]: Why did you delete this constraint?

File: ConsumerApi/RelationshipChangesDbContextSeeder.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    public RelationshipChangesDbContextSeeder(IServiceProvider serviceProvider)
+    {
+        _blobStorage = serviceProvider.GetService<IBlobStorage>();
+        _blobRootFolder = serviceProvider.GetService<IOptions<BlobOptions>>()!.Value.RootFolder;
+    }
+
+    public async Task SeedAsync(RelationshipsDbContext context)
+    {
+        if (_blobRootFolder == null)
```
Comment (tnotheis) [inline ConsumerApi/RelationshipChangesDbContextSeeder.cs:29]: Can you add a comment here that explains why we return in this case? Because I bet we will forget it. ;)

File: ConsumerApi/RelationshipChangesDbContextSeeder.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            return;
+
+        var relationshipChanges = await context.RelationshipChanges
```
Comment (tnotheis) [inline ConsumerApi/RelationshipChangesDbContextSeeder.cs:32]:
```suggestion
        var relationshipChangesWithMissingContents = await context.RelationshipChanges
```

File: ConsumerApi/RelationshipChangesDbContextSeeder.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                    context.RelationshipChanges.Update(relationshipChange);
+                }
+                catch (NotFoundException)
+                {
+                    // response content is optional so we can ignore this
+                }
+            }
+
+            await context.SaveChangesAsync();
+        }
```
Comment (tnotheis) [inline ConsumerApi/RelationshipChangesDbContextSeeder.cs:57]: Extract all of this into a method `FillContentColumnsFromBlobStorage`.

File: ConsumerApi/RelationshipChangesDbContextSeeder.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        if (_blobRootFolder == null)
+            return;
+
+        var relationshipChanges = await context.RelationshipChanges
+            .Where(rc =>  rc.Request.Content == null || (rc.Response != null && rc.Response.Content == null)).ToListAsync();
+
+        foreach (var relationshipChange in relationshipChanges)
+        {
+            if (relationshipChange.Request != null && relationshipChange.Request.Content == null)
```
Comment (tnotheis) [inline ConsumerApi/RelationshipChangesDbContextSeeder.cs:37]: Why do you check for `relationshipChange.Request != null`? The `Request` property should never be null.

File: ConsumerApi/RelationshipChangesDbContextSeeder.cs (hunk 2 above)
Comment (tnotheis) [inline ConsumerApi/RelationshipChangesDbContextSeeder.cs:29]: Maybe we should also check if _blobStorage is null. This way we can get rid of the `!`s below.

```suggestion
        if (_blobRootFolder == null || _blobStorage == null)
```

File: ConsumerApi/RelationshipChangesDbContextSeeder.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            {
+                relationshipChange.Request.LoadContent(await _blobStorage!.FindAsync(_blobRootFolder, relationshipChange.Request.Id));
```
Comment (tnotheis) [inline ConsumerApi/RelationshipChangesDbContextSeeder.cs:39]: You should extract the blob content to a variable to make it more readable. Same for the response

File: Modules/Relationships/src/Relationships.ConsumerApi/RelationshipsModule.cs [hunk 7]
```diff
@@ -29,12 +29,17 @@ public override void ConfigureServices(IServiceCollection services, IConfigurati
...
+            if (parsedConfiguration.Infrastructure.BlobStorage != null)
+            {
+                options.BlobStorageOptions = new();
+
+                options.BlobStorageOptions.CloudProvider = parsedConfiguration.Infrastructure.BlobStorage.CloudProvider;
+                options.BlobStorageOptions.ConnectionInfo = parsedConfiguration.Infrastructure.BlobStorage.ConnectionInfo;
+                options.BlobStorageOptions.Container =
+                    parsedConfiguration.Infrastructure.BlobStorage.ContainerName.IsNullOrEmpty()
+                        ? "relationships"
+                        : parsedConfiguration.Infrastructure.BlobStorage.ContainerName;
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.ConsumerApi/RelationshipsModule.cs:41]:
```suggestion
                options.BlobStorageOptions = new BlobStorageOptions {
                    CloudProvider = parsedConfiguration.Infrastructure.BlobStorage.CloudProvider,
                    ConnectionInfo = parsedConfiguration.Infrastructure.BlobStorage.ConnectionInfo,
                    Container =
                        parsedConfiguration.Infrastructure.BlobStorage.ContainerName.IsNullOrEmpty()
                            ? "relationships"
                            : parsedConfiguration.Infrastructure.BlobStorage.ContainerName
                };
```

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs [hunk 8]
```diff
@@ -67,42 +60,30 @@ public class RelationshipsRepository : IRelationshipsRepository
...
         var changes = await query.OrderAndPaginate(d => d.CreatedAt, paginationFilter, cancellationToken);
 
-        await FillContentOfChanges(changes.ItemsOnPage);
-
         return changes;
     }
 
     public async Task<Relationship> FindRelationship(RelationshipId id, IdentityAddress identityAddress,
-        CancellationToken cancellationToken, bool track = false, bool fillContent = true)
+        CancellationToken cancellationToken, bool track = false, bool fillContent = true) // todo: fill content is redundant
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:67]: No todo comments

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs [hunk 9]
```diff
@@ -67,42 +60,30 @@ public class RelationshipsRepository : IRelationshipsRepository
...
-            await FillContentOfChanges(relationship.Changes);
-        }
-
         return relationship;
     }
 
     public async Task<RelationshipChange> FindRelationshipChange(RelationshipChangeId id,
         IdentityAddress identityAddress, CancellationToken cancellationToken, bool track = false,
-        bool fillContent = true)
+        bool fillContent = true) // todo: fillContent is redundant
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:79]: No todo comments

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs [hunk 10]
```diff
@@ -117,23 +98,21 @@ public class RelationshipsRepository : IRelationshipsRepository
...
-        await FillContentOfChanges(relationships.ItemsOnPage.SelectMany(r => r.Changes));
-
         return relationships;
     }
 
     public async Task Update(Relationship relationship)
     {
         _relationships.Update(relationship);
-        await SaveContentOfLatestChange(relationship);
+        // await SaveContentOfLatestChange(relationship);
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:107]: No commented out code

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs [hunk 11]
```diff
@@ -117,23 +98,21 @@ public class RelationshipsRepository : IRelationshipsRepository
...
         await _dbContext.SaveChangesAsync();
     }
 
 
     public async Task Add(Relationship relationship, CancellationToken cancellationToken)
     {
         await _relationships.AddAsync(relationship, cancellationToken);
-        await SaveContentOfLatestChange(relationship);
+        // await SaveContentOfLatestChange(relationship);
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:115]: No commented out code

File: Modules/Relationships/test/Relationships.Application.Tests/RelationshipRepositoryTests.cs
Comment (tnotheis) [inline Modules/Relationships/test/Relationships.Application.Tests/RelationshipRepositoryTests.cs:1]: Why do these tests still exist?

File: Modules/Relationships/test/Relationships.Application.Tests/Relationships.Application.Tests.csproj [hunk 12]
```diff
@@ -16,9 +16,9 @@
 	</ItemGroup>
 
 	<ItemGroup>
-		<ProjectReference Include="..\..\..\..\BuildingBlocks\src\UnitTestTools\UnitTestTools.csproj" />
 		<ProjectReference Include="..\..\src\Relationships.Application\Relationships.Application.csproj" />
 		<ProjectReference Include="..\..\src\Relationships.Infrastructure\Relationships.Infrastructure.csproj" />
+        <ProjectReference Include="..\..\..\Devices\test\Devices.Application.Tests\Devices.Application.Tests.csproj" />
```
Comment (tnotheis) [inline Modules/Relationships/test/Relationships.Application.Tests/Relationships.Application.Tests.csproj:21]: Again, there's this reference to the Devices.Application.Tests project...

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: BuildingBlocks/src/BuildingBlocks.API/Extensions/HostExtensions.cs (hunk 1 above)
Comment (NikolaVetnic) [reply BuildingBlocks/src/BuildingBlocks.API/Extensions/HostExtensions.cs:72]: No idea, I don't even remember touching the file or seeing it in the changes. My bad & reverted.

File: Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs (hunk 8 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:67]: Comment removed, but should we remove the input parameter?

Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:67]: Sure.

Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs]: Oh, wrong place

Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs]: Ähm, not the wrong place. :) So yeah, you should still remove the parameter.

**Review by tnotheis — CHANGES_REQUESTED** [review]
See unresolved comment

**Comment by NikolaVetnic** [comment]
Removed the `fillContent` input parameter as well as all the methods and method calls that had anything to do with filling the `RelationshipChangeRequest.Content` or `RelationshipChangeResponse.Content`.

**Review by tnotheis — CHANGES_REQUESTED** [review]
Now you added the column, but you don't fill it in the seeder

**Comment by NikolaVetnic** [comment]
> Now you added the column, but you don't fill it in the seeder

I don't get it, I have this code - isn't it supposed to seed the database:
```csharp
        foreach (var relationshipChange in relationshipChangesWithMissingContents)
        {
            if (relationshipChange.Request.Content == null)
            {
                var blobRequestContent = await _blobStorage.FindAsync(_blobRootFolder, relationshipChange.Request.Id);
                relationshipChange.Request.LoadContent(blobRequestContent);
                context.RelationshipChanges.Update(relationshipChange);
            }

            if (relationshipChange.Response is { Content: null })
            {
                try
                {
                    var blobResponseContent = await _blobStorage.FindAsync(_blobRootFolder, relationshipChange.Response.Id);
                    relationshipChange.Response.LoadContent(blobResponseContent);
                    context.RelationshipChanges.Update(relationshipChange);
                }
                catch (NotFoundException)
                {
                    // response content is optional so we can ignore this
                }
            }

            await context.SaveChangesAsync();
        }
```

That's the same thing we had with Tokens and Messages.

**Comment by tnotheis** [comment]
> > Now you added the column, but you don't fill it in the seeder
> 
> I don't get it, I have this code - isn't it supposed to seed the database:
> 
> ```cs
>         foreach (var relationshipChange in relationshipChangesWithMissingContents)
>         {
>             if (relationshipChange.Request.Content == null)
>             {
>                 var blobRequestContent = await _blobStorage.FindAsync(_blobRootFolder, relationshipChange.Request.Id);
>                 relationshipChange.Request.LoadContent(blobRequestContent);
>                 context.RelationshipChanges.Update(relationshipChange);
>             }
> 
>             if (relationshipChange.Response is { Content: null })
>             {
>                 try
>                 {
>                     var blobResponseContent = await _blobStorage.FindAsync(_blobRootFolder, relationshipChange.Response.Id);
>                     relationshipChange.Response.LoadContent(blobResponseContent);
>                     context.RelationshipChanges.Update(relationshipChange);
>                 }
>                 catch (NotFoundException)
>                 {
>                     // response content is optional so we can ignore this
>                 }
>             }
> 
>             await context.SaveChangesAsync();
>         }
> ```
> 
> That's the same thing we had with Tokens and Messages.

Yes, it does it for the Relationship Changes. But not for the Templates

File: ConsumerApi/RelationshipsDbContextSeeder.cs [hunk 13]
```diff
@@ -64,4 +68,22 @@ private async Task FillContentColumnsFromBlobStorage(RelationshipsDbContext cont
...
     }
+
+    private async Task FillRelationshipTemplatesWithMissingContent(RelationshipsDbContext context)
+    {
+        var relationshipTemplatesWithMissingContent = await context.RelationshipTemplates
+            .Where(rt => rt.Content == null).ToListAsync();
+
+        foreach (var relationshipTemplate in relationshipTemplatesWithMissingContent)
+        {
+            if (relationshipTemplate.Content == null)
```
Comment (tnotheis) [inline ConsumerApi/RelationshipsDbContextSeeder.cs:79]: Why do you check if the content is null? You only selected those templates which don't have a content.

**Review by tnotheis — CHANGES_REQUESTED** [review]
Remember to request review once you're done with all changes.

File: ConsumerApi/RelationshipsDbContextSeeder.cs (hunk 13 above)
Comment (NikolaVetnic) [reply ConsumerApi/RelationshipsDbContextSeeder.cs:79]: Stupid mistake :/

**Comment by tnotheis** [comment]
The tests now succeed. But you have some conflicts with main, which you should resolve.

**Review by tnotheis — CHANGES_REQUESTED** [review]
Please have a look at my comment before you request review

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: ConsumerApi/RelationshipsDbContextSeeder.cs [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+                    relationshipChange.Response.LoadContent(blobResponseContent);
+                    context.RelationshipChanges.Update(relationshipChange);
+                }
+                catch (NotFoundException)
+                {
+                    // response content is optional so we can ignore this
+                }
+            }
+
+            await context.SaveChangesAsync();
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply ConsumerApi/RelationshipsDbContextSeeder.cs:68]: Messages and Tokens were fine already.

File: ConsumerApi/RelationshipsDbContextSeeder.cs [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        var relationshipTemplatesWithMissingContent = await context.RelationshipTemplates
+            .Where(rt => rt.Content == null).ToListAsync();
+
+        foreach (var relationshipTemplate in relationshipTemplatesWithMissingContent)
+        {
+            var blobContent = await _blobStorage!.FindAsync(_blobRootFolder!, relationshipTemplate.Id);
+            relationshipTemplate.LoadContent(blobContent);
+            context.RelationshipTemplates.Update(relationshipTemplate);
+
+            await context.SaveChangesAsync();
```
Comment (tnotheis) [inline ConsumerApi/RelationshipsDbContextSeeder.cs:83]: ...

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — APPROVED** [review]
